### ISSUE TITLE
Description of a manual installation is missing

### ISSUE THREAD

**Issue by weierophinney** [mezzio/mezzio#8]
I'm trying to include zend-expressive in my project, but unsuccessfully.
When i require zend-expressive via console ( composer require zendframework/zend-expressive ), i get the following error: 

```
Your requirements could not be resolved to an installable set of packages.
  Problem 1
    - zendframework/zend-expressive 3.0.0 requires zendframework/zend-httphandlerrunner ^1.0.1 -> satisfiable by zendframework/zend-httphandlerrunner[1.0.1].
    - zendframework/zend-expressive 3.0.1 requires zendframework/zend-httphandlerrunner ^1.0.1 -> satisfiable by zendframework/zend-httphandlerrunner[1.0.1].
    - zendframework/zend-expressive 3.0.2 requires zendframework/zend-httphandlerrunner ^1.0.1 -> satisfiable by zendframework/zend-httphandlerrunner[1.0.1].
    - zendframework/zend-httphandlerrunner 1.0.1 requires psr/http-message-implementation ^1.0 -> no matching package found.
    - Installation request for zendframework/zend-expressive ^3.0 -> satisfiable by zendframework/zend-expressive[3.0.0, 3.0.1, 3.0.2].

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.
```

### Code to reproduce the issue

```
composer require zendframework/zend-expressive
```

### Expected results
Expected composer to install the package

### Actual results
Composer returned an error regarding package dependencies

---

Originally posted by @itapai at https://github.com/zendframework/zend-expressive/issues/624

**Comment by weierophinney** [comment]
It's not broken, you are missing a `psr/http-message-implementation`, e.g. `zendframework/zend-diactoros`

So I guess this would do:
`composer require zendframework/zend-expressive zendframework/zend-diactoros`

---

Originally posted by @JoeBengalen at https://github.com/zendframework/zend-expressive/issues/624#issuecomment-403813033

**Comment by weierophinney** [comment]
@ksemeks One new tool for you: `composer why-not <package>`. This will give you detailed information on what blocks installation of a package.

That said, your error message indicates the issue:

```text
zendframework/zend-httphandlerrunner 1.0.1 requires psr/http-message-implementation ^1.0 -> no matching package found.
```

We do not currently cover installation within an existing project, which tells me this should be documented. I'm adding the "documentation" and "help wanted" labels presently, so somebody can pick that up.

---

Originally posted by @weierophinney at https://github.com/zendframework/zend-expressive/issues/624#issuecomment-403832747

**Comment by arueckauer** [comment]
Currently Mezzio requires an implementation of `psr/http-message-implementation` as well as `laminas/laminas-httphandlerrunner`. This can be achieved with e.g.:

```bash
composer require laminas/laminas-diactoros laminas/laminas-httphandlerrunner mezzio/mezzio
```

However, this does not add the same value as utilizing `mezzio/mezzio-skeleton` which in addition sets up a directory structure and installs a DI, a router and optionally a template engine and error handler.

The documentation of this process is cumbersome and of little value, since it first adds little value for very few users using this scenario and secondly it is much easier to run the `mezzio/mezzio-skeleton` installer and copy the necessary components over to the existing project.

IMO this is a won't fix of an edge case.

**Comment by froschdesign** [comment]
@arueckauer 
> The documentation of this process is cumbersome and of little value, since it first adds little value for very few users using this scenario and secondly it is much easier to run the `mezzio/mezzio-skeleton` installer and copy the necessary components over to the existing project.

A documentation page for the installation is still missing and a _manual_ installation should also be added there.



> ```shell
> composer require laminas/laminas-diactoros laminas/laminas-httphandlerrunner mezzio/mezzio
> ```

laminas-httphandlerrunner can be omitted from this command:

https://github.com/mezzio/mezzio/blob/d5bf81f0caee6cf4b682a4bf1f8d693288bf217d/composer.json#L39-L43

**Comment by arueckauer** [comment]
> A documentation page for the installation is still missing and a _manual_ installation should also be added there.

I am missing something here. What is the value of adding only `mezzio/mezzio` to an existing project? Would the documentation cover routing and DI, application structure, ...?

> > ```shell
> > composer require laminas/laminas-diactoros laminas/laminas-httphandlerrunner mezzio/mezzio
> > ```
> 
> laminas-httphandlerrunner can be omitted from this command:
> 
> https://github.com/mezzio/mezzio/blob/d5bf81f0caee6cf4b682a4bf1f8d693288bf217d/composer.json#L39-L43

You are right. Good catch!

**Comment by froschdesign** [comment]
> I am missing something here. What is the value of adding only `mezzio/mezzio` to an existing project? Would the documentation cover routing and DI, application structure, ...?

Sorry, I was a bit unclear here. I mean only the manual installation, not any integration. I think an integration or a migration can be a topic for blog post or something similar.

**Comment by arueckauer** [comment]
So, adding this to documentation would suffice?

> Prior to manually installing Mezzio, check if the project has an implementation of [`psr/http-message-implementation`](https://packagist.org/providers/psr/http-message-implementation). If does not, add one of your choice, e.g.:
> ```bash
> composer require laminas/laminas-diactoros mezzio/mezzio
> ```

**Comment by froschdesign** [comment]
> So, adding this to documentation would suffice?

We need a separate page which shows the different ways for an installation. Creating a project via the skeleton application is one of them. The _manual_ installation may be another.
An entry "Installation" is completely missing in the navigation: https://docs.mezzio.dev/mezzio/

**Comment by froschdesign** [comment]
Will be fixed with #119
